### PR TITLE
thchang / 458 rename layout

### LINKED ISSUE #458: rename saved layouts
I made a mistake naming a layout and want to rename it.  There doesn't seem to be a way to rename layouts, only to delete or save.  This is a minor issue as one can always delete and re save.  But it would be slick to be able to rename saved layout.

### PR DESCRIPTION
**Pull request by TienHao**
**Description**

closes #458. Added the rename button in the save layout dialog which users can rename layout name directly.

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] e2e test passing / ~corresponding fix added~
- [x] changelog updated / ~no changelog update needed~
- [x] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [x] `BackendService` unchanged / ~`BackendService` changed and corresponding ICD test fix added~

### CODE REVIEW

**Comment by kswang1029** [comment]
@TienHao overall this looks good. 👍 There are few things to be considered:
1) if we rename a layout with a name same as other existing layouts, a popup will show up as a warning. Then it disappeared. I suggest once the user clicked the ok button, we show the rename pop up again so users can try another name.
2) please add a check for unix/linux reserved characters and words and prevent users using those as part of the layout name. They are (https://www.cyberciti.biz/faq/linuxunix-rules-for-naming-file-and-directory-names/)
Screen Shot 2022-10-28 at 11 37 20

Please help me to check if there are other forbidden characters.

I tried `>` and terrible things happened if I try to open it with vi in the terminal. 🤦🏻‍♂️ Also tried a space and it works but it is weird to have a space as the layout's name/json file name.

**Comment by TienHao** [comment]
@kswang1029

 I have fixed the first item.

About the space character in second item, I think it is OK to have that since we have space characters in several preset layout like "Cube View", "Cube Analysis". Or, we replace all space characters in preset layouts once a space is not allowed in a layout name. either way is ok for me.

I added a alert tooltips to be shown when a forbidden character has be entered.截圖 2022-11-30 下午3 01 50

**Comment by kswang1029** [comment]
> @kswang1029
> 
> I have fixed the first item.
> 
> About the space character in second item, I think it is OK to have that since we have space characters in several preset layout like "Cube View", "Cube Analysis". Or, we replace all space characters in preset layouts once a space is not allowed in a layout name. either way is ok for me.
> 
> I added a alert tooltips to be shown when a forbidden character has be entered.截圖 2022-11-30 下午3 01 50

ok to have space in the names but we should not allow a name with a single space only.

**Comment by kswang1029** [comment]
@TienHao I tried `\` so the saved file is `\.json`. Odd thing happens when I try to use vi to open it. So I think we will need to include `\` as forbidden char.  `;` needs to be included too as if we do `vi ;.json` the file is not loaded.

**Comment by kswang1029** [comment]
@TienHao on a related issue: in fact when we try to "save" a layout, we should also check special char. Could you also fix it too in the same PR?

**Comment by TienHao** [comment]
@kswang1029 

I have added `\` and `;` into the forbidden char set. The special char check in the "save" layout has already been implemented.

**Comment by kswang1029** [comment]
> @kswang1029
> 
> I have added `\` and `;` into the forbidden char set. The special char check in the "save" layout has already been implemented.

@TienHao I tried all possible special char and loaded the file with `vi` and compiled the following list to see which should be forbidden (due to zsh/bash semantics). Please include those with "no" to the forbidden list. 

```
Char	canUse?
========================
~	no
`	no
!	no
@	yes
#	yes
$	yes
%	yes
^	yes
&	no
*	no
(	no
)	no
-	no
_       yes
+	no
=	no
[	no	
]	yes
{	yes
}	yes
\	no
|	no
;	no
:	no
'	no
<	no
>	no
.	no
,	yes
?	no
/	no 
```

File: src/stores/LayoutStore.ts [hunk 1]
```diff
@@ -185,6 +195,64 @@ export class LayoutStore {
...
 
+    @flow.bound *renameLayout(oldName: string, newName: string) {
+        const appStore = AppStore.Instance;
+
+        if (!this.layouts || !newName || !this.dockedLayout) {
+            appStore.alertStore.showAlert("Save layout failed! Empty layouts or name.");
+            return;
+        }
+
+        if (appStore.layoutStore.layoutExists(newName)) {
```
Comment (YuHsuan-Hwang) [inline src/stores/LayoutStore.ts]: can be simplified to `this.layoutExists`

File: src/stores/LayoutStore.ts [hunk 2]
```diff
@@ -185,6 +195,64 @@ export class LayoutStore {
...
+            appStore.alertStore.showAlert("Layout name already exists.");
+            return;
+        }
+
```
Comment (YuHsuan-Hwang) [inline src/stores/LayoutStore.ts]: Shall we also check for preset names as in `saveLayout`?

File: src/stores/LayoutStore.ts [hunk 3]
```diff
@@ -185,6 +195,64 @@ export class LayoutStore {
...
+        appStore.dialogStore.hideSaveLayoutDialog();
+
+        // save layout to layouts[] & server/local storage
+        const configToSave = this.layouts[oldName];
+        this.layouts[newName] = configToSave;
+        if (!PresetLayout.isPreset(this.layoutNameToBeSaved)) {
+            try {
+                const success = yield appStore.apiService.setLayout(newName, configToSave);
+                if (success) {
+                    appStore.apiService.clearLayout(oldName).then(
```
Comment (YuHsuan-Hwang) [inline src/stores/LayoutStore.ts]: Could we use `yield` here? Maybe also adjust `deleteLayout` to use `@flow.bound` and `yield`.

**Review by YuHsuan-Hwang — CHANGES_REQUESTED** [review]

File: src/stores/LayoutStore.ts (hunk 3 above)
Comment (YuHsuan-Hwang) [reply src/stores/LayoutStore.ts]: @TienHao Maybe we can remove `then()` by
```
try {
    const success = yield appStore.apiService.setLayout(newName, configToSave);
    if (success) {
        const success = yield appStore.apiService.clearLayout(oldName);
        if (success) {
            delete this.layouts[oldName];
        }
        this.handleRenameResult(oldName, newName, success);
    }
} catch (err) {
    console.log(err);
    this.handleRenameResult(oldName, newName, false);
}
```
We can also try to remove `then()` in `deleteLayout` similarly.

File: src/components/Dialogs/LayoutDialog/SaveLayoutDialogComponent.tsx [hunk 4]
```diff
@@ -54,13 +54,24 @@ export class SaveLayoutDialogComponent extends React.Component {
...
+        this.clearInput();
+    };
+
     @computed get isEmpty(): boolean {
-        return !this.layoutName;
+        return !this.layoutName?.trim();
+    }
+
+    @computed get validName(): boolean {
+        return this.layoutName.match(/^[^~`!*()-+=[.'?<>/|\\:;&]+$/)?.length > 0;
```
Comment (kswang1029) [inline src/components/Dialogs/LayoutDialog/SaveLayoutDialogComponent.tsx]: @TienHao could you check why the minus sign "-" is _not_ detected as invalid char?

File: src/stores/LayoutStore/LayoutStore.ts [hunk 5]
```diff
@@ -186,30 +196,81 @@ export class LayoutStore {
...
+        const appStore = AppStore.Instance;
+        if (!layoutName || !this.layoutExists(layoutName)) {
+            appStore.alertStore.showAlert(`Cannot delete layout ${layoutName}! It does not exist.`);
+            return;
+        }
+
+        try {
+            const success = yield appStore.apiService.clearLayout(layoutName);
+            if (success) {
+                delete this.layouts[layoutName];
```
Comment (YuHsuan-Hwang) [inline src/stores/LayoutStore/LayoutStore.ts:265]: The new changes look good! Just one minor question: do we need to preserve the lines `if (layoutName === this.currentLayoutName) {  this.currentLayoutName = ""; }` here?

Comment (TienHao) [reply src/stores/LayoutStore/LayoutStore.ts:265]: I am thinking that these lines still need to be preserved. If a user delete a current layout (eg. _layout_A_) without clear `this.currentLayoutName`, and then rename another layout(eg. _layout_B_) or add a new layout with the deleted name (_layout_A_), the current layout hint will be misleading.

**Review by kswang1029 — APPROVED** [review]
looks good. please proceed code review

File: src/stores/LayoutStore/LayoutStore.ts (hunk 5 above)
Comment (YuHsuan-Hwang) [reply src/stores/LayoutStore/LayoutStore.ts:265]: Oh, I see we also have them in `handleDeleteResult`. I missed that.

**Review by YuHsuan-Hwang — APPROVED** [review]

**Pull request merged**